File: src/main/java/jukebot/commands/Unqueue.java
package jukebot.commands;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import jukebot.audioutilities.AudioHandler;
import jukebot.utils.Command;
import jukebot.utils.CommandProperties;
import jukebot.utils.Context;
import jukebot.utils.Helpers;

@CommandProperties(description = "Remove a track from the queue", aliases = {"uq", "remove", "r"}, category = CommandProperties.category.MEDIA)
public class Unqueue implements Command {

    public void execute(final Context context) {

        final AudioHandler player = context.getAudioPlayer();

        if (player.getQueue().isEmpty()) {
            context.embed("Nothing to Unqueue", "The queue is empty!");
            return;
        }

        if (context.getArgString().isEmpty()) {
            context.embed("Specify track index", "You need to specify the index of the track to unqueue.");
            return;
        }

        final int selected = Helpers.Companion.parseNumber(context.getArgString(), 0);

        if (selected < 1 || selected > player.getQueue().size()) {
            context.embed("Invalid position specified!", "You need to specify a valid target track.");
            return;
        }

        final AudioTrack selectedTrack = player.getQueue().get(selected - 1);

        if ((long) selectedTrack.getUserData() != context.getAuthor().getIdLong() && !context.isDJ(false)) {
            context.embed("Not a DJ", "You need the DJ role to unqueue others' tracks. [See here on how to become a DJ](https://jukebot.serux.pro/faq)");
            return;
        }

        player.getQueue().remove(selected - 1);

        context.embed("Track Unqueued", "Removed **" + selectedTrack.getInfo().title + "** from the queue.");

    }
}

File: src/main/java/jukebot/commands/Unqueue.kt
package jukebot.commands

import jukebot.framework.Command
import jukebot.framework.CommandCategory
import jukebot.framework.CommandProperties
import jukebot.framework.Context

@CommandProperties(description = "Remove a track from the queue", aliases = ["uq", "remove", "r"], category = CommandCategory.QUEUE)
class Unqueue : Command(ExecutionType.STANDARD) {

    override fun execute(context: Context) {
        val player = context.getAudioPlayer()

        if (player.queue.isEmpty()) {
            return context.embed("Queue Empty", "There are no tracks to remove.")
        }

        if (context.argString.isEmpty()) {
            return context.embed("Specify track index", "You need to specify the index of the track to remove.")
        }

        // TODO: Accept a range? i.e. 3-7, and/or multiple numbers, i.e. 3 8 11 12 21
        val selected = context.args.firstOrNull()?.toIntOrNull() ?: 0

        if (selected < 1 || selected > player.queue.size) {
            return context.embed("Invalid position specified!", "You need to specify a valid target track.")
        }

        val selectedTrack = player.queue[selected - 1]

        if (selectedTrack.userData as Long != context.author.idLong && !context.isDJ(false)) {
            return context.embed("Not a DJ", "You need the DJ role to unqueue others' tracks. [See here on how to become a DJ](https://jukebot.serux.pro/faq)")
        }

        player.queue.removeAt(selected - 1)
        context.embed("Track Unqueued", "Removed **${selectedTrack.info.title}** from the queue.")
    }
}
